### PR TITLE
feat: add KV cache CPU backup for sleep/wakeup radix tree preservation

### PR DESCRIPTION
**Pull request by tongguang666**
Add --enable-kv-cache-cpu-backup flag that preserves the prefix cache
    radix tree across sleep/wakeup cycles.

    Changes:
    - server_args.py: new --enable-kv-cache-cpu-backup flag with validation
    - radix_cache.py: export_snapshot() and import_snapshot() methods (iterative DFS to avoid recursion limit on large trees)
    - scheduler_update_weights_mixin.py: snapshot before pause(), restore after resume(); pre-flight CPU memory check (hard reject at 1.2x safety margin, soft warning at 1.5x); graceful exit on insufficient memory via ReleaseMemoryOccupationReqOutput(success=False, message=...)
    - memory_pool.py: enable_kv_cpu_backup param threaded through KVCache and all subclasses (MHA, MLA, NSA, DoubleSparse, HybridLinear, HiSparseNSA)
    - model_runner_kv_cache_mixin.py: pass enable_kv_cpu_backup in all pool creation calls
    - io_struct.py: ReleaseMemoryOccupationReqOutput gains success/message fields for graceful failure reporting
    - test/srt/test_kv_cache_cpu_backup.py: complete test suite (313 lines)

## Motivation

During sleep/wakeup cycles (e.g., `release_memory_occupation` / `resume_memory_occupation` for weight updates like RLHF), the KV cache radix tree is flushed, causing prefix cache hits to be lost after wakeup. This forces full prefill for all subsequent requests, degrading performance.

This PR adds `--enable-kv-cache-cpu-backup` which preserves the prefix cache radix tree across sleep/wakeup cycles by serializing it before `pause()` and restoring it after `resume()`, enabling prefix cache hits to work immediately after wakeup without re-prefill.

### Core feature
- **`radix_cache.py`**: Added `export_snapshot()` (iterative DFS to avoid recursion limits on deep trees) and `import_snapshot()` methods for serializing/deserializing the radix tree topology and slot assignments. Comprehensive error handling with descriptive messages for node rebuild failures, topology wiring errors, and allocator free-list restoration failures.
- **`scheduler_update_weights_mixin.py`**: On `release_memory_occupation`, snapshots the radix tree before `pause()`. On `resume_memory_occupation`, restores it after `resume()`. Also added a pre-flight CPU memory check: hard reject at 1.2x safety margin, soft warning at 1.5x, preventing `cudaMallocHost` OOM kills.
- **`io_struct.py`**: `ReleaseMemoryOccupationReqOutput` gains `success: bool` and `message: str` fields for graceful failure reporting.
- **`memory_pool.py`**: Threaded `enable_kv_cpu_backup` through `KVCache` base class and all subclasses (`MHATokenToKVPool`, `MHATokenToKVPoolFP4`, `MLATokenToKVPool`, `MLATokenToKVPoolFP4`, `NSATokenToKVPool`, `DoubleSparseTokenToKVPool`, `HybridLinearKVPool`, `HybridReqToTokenPool`). Each `_create_buffers()` passes `enable_cpu_backup=self.enable_kv_cpu_backup` to `memory_saver_adapter.region()`.
- **`hisparse_memory_pool.py`**: `HiSparseNSATokenToKVPool` passes `enable_kv_cpu_backup` to parent.
- **`model_runner_kv_cache_mixin.py`**: Propagates `enable_kv_cpu_backup` to all KV pool creation calls across all cache backends.
- **`server_args.py`**: Added `--enable-kv-cache-cpu-backup` flag with validation: requires `--enable-memory-saver` and is incompatible with `--disable-radix-cache`.
- **Container-aware memory check**: `_get_container_available_memory_bytes()` reads cgroup v2/v1 memory limits first (correct for Docker/K8s), falling back to `psutil.virtual_memory().available` for non-containerized hosts.

### Tests
- **`test/srt/test_kv_cache_cpu_backup.py`**: Complete test suite (313 lines) covering:
  - Unit tests: empty tree export, snapshot round-trip, prefix match after restore, allocator free-list update, disabled/empty snapshot handling.
  - Integration tests: basic sleep/wakeup correctness (output consistency, GPU memory drop/recovery), prefix cache restoration (cached_tokens > 0 after wakeup).

## Accuracy Tests

- **Output consistency**: Verified that generation output before sleep matches output after wakeup across multiple cycles (TP=1 and TP=2).
- **Prefix cache restoration**: Verified that `cached_tokens > 0` after wakeup, confirming the radix tree snapshot is correctly restored and prefix cache hits fire.
- **Radix tree integrity**: Unit tests confirm `total_size()` and `match_prefix()` results are identical before/after snapshot round-trip.

### Sleep/Wakeup Timing (Qwen3-32B, NVIDIA A100, TP=1)

| Metric | First cycle | Subsequent cycles |
|--------|------------|-------------------|
| Sleep (KV only) | ~2.1s | ~0.9s |
| Wakeup (KV only) | ~0.36s | ~0.35s |
| Sleep (KV + Weight) | ~24s | ~9.5s |
| Wakeup (KV + Weight) | ~3.85s | ~3.85s |

### Sleep/Wakeup Timing (Qwen3-32B, NVIDIA A100, TP=2)

| Metric | Value |
|--------|-------|
| Sleep (KV + Weight) | ~32.5s |
| Wakeup (KV + Weight) | ~3.9s |

### Stability
- 10× consecutive sleep/wakeup cycle test passed with no crashes, no memory leaks, and consistent generation output across all cycles.
- Pre-flight memory check correctly rejects release requests when available CPU memory is below 1.2x the required pinned memory size, preventing process kills from `cudaMallocHost` failures

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

File: python/sglang/srt/mem_cache/radix_cache.py [hunk 1]
```diff
@@ -904,6 +904,218 @@ def take_events(self):
...
+                        device=self.device,
+                    )
+                    used_mask = torch.zeros(
+                        allocator.size + 1, dtype=torch.bool, device=self.device
+                    )
+                    used_mask[used_tensor] = True
+                    allocator.free_pages = all_slots[~used_mask[1:]]
+                    allocator.release_pages = torch.empty(
+                        (0,), dtype=torch.int64, device=self.device
+                    )
```
[the remark this answers is not in the document]
Comment (tongguang666) [reply python/sglang/srt/mem_cache/radix_cache.py:1105]: Thanks for the review! However, this concern is based on a misunderstanding of the allocator's indexing scheme.

SGLang's KVCache allocators use 1-based indexing, not 0-based. From python/sglang/srt/mem_cache/allocator.py line 133:

Copy
# The padded slot 0 is used for writing dummy outputs from padded tokens.
self.free_pages = torch.arange(1, self.size + 1, dtype=torch.int64, device=self.device)
The comment explicitly states that slot 0 is a dummy padding slot for padded tokens. All valid page indices are [1, 2, ..., size].

Therefore our restoration code is correct:

if slot > 0: skips the dummy padding slot (correct)
torch.arange(1, allocator.size + 1) generates [1, 2, ..., size] (correct)
used_mask = torch.zeros(allocator.size + 1) with ~used_mask[1:] properly excludes slot 0 from free_pages (correct)

File: python/sglang/srt/managers/scheduler_update_weights_mixin.py [hunk 2]
```diff
@@ -136,9 +189,90 @@ def release_memory_occupation(
...
+                    else None
+                )
+                logger.info(
+                    "KV cache CPU backup enabled: saving radix tree snapshot before sleep."
+                )
+                # pause() triggers GPU→CPU copy via torch_memory_saver cpu_backup path
+                self.memory_saver_adapter.pause(GPU_MEMORY_TYPE_KV_CACHE)
+                # Flush only the metadata (free-list and radix tree) without touching the
+                # already-paused GPU tensors.  We do NOT call self.flush_cache() here
+                # because that would clear the slot assignments we just snapshotted.
```
[the remark this answers is not in the document]
Comment (tongguang666) [reply python/sglang/srt/managers/scheduler_update_weights_mixin.py:274]: Thanks for the review! While the radix tree Python objects are indeed preserved across pause/resume, the `import_snapshot()` logic serves a critical purpose beyond just restoring the tree:

**It explicitly rebuilds the allocator's `free_pages` to match the radix tree's slot assignments.**

Even though `flush_cache()` is skipped, the allocator state and radix tree are independent data structures. If they ever get out of sync (e.g., due to edge cases during pause/resume or partial failures), we'd have a double-allocation bug where pages referenced by the radix tree are re-allocated to new requests.

The `import_snapshot()` call ensures:
1. `allocator.free_pages` explicitly excludes all pages currently used by the radix tree
2. The allocator state is deterministically consistent with the tree after restore

The CPU overhead is minimal compared to the safety guarantee. We prefer defensive correctness over premature optimization here.

File: python/sglang/srt/mem_cache/radix_cache.py [hunk 3]
```diff
@@ -904,6 +904,218 @@ def take_events(self):
...
+                node.lock_ref = nd["lock_ref"]
+                node.key = RadixKey(
+                    token_ids=nd["token_ids"], extra_key=nd["extra_key"]
+                )
+                if nd["value"]:
+                    node.value = torch.tensor(
+                        nd["value"], dtype=torch.int64, device=self.device
+                    )
+                else:
+                    node.value = []
```
[the remark this answers is not in the document]
Comment (tongguang666) [reply python/sglang/srt/mem_cache/radix_cache.py]: accept

**Comment by tongguang666** [comment]
"Hi @ispobock @xiezhq-hermann, could you please help to run /tag-run-ci-label to trigger CI? Thanks!"

**Comment by hnyls2002** [comment]
Hi @tongguang666, thanks for the contribution. This has been open ~3 months without maintainer review and is now conflicting with main; the sleep/wakeup radix preservation it implements would need to be reworked against the current scheduler/cache code. Closing for inactivity - please reopen if you'd like to rebase and push it forward.
